internal/v4/auth.go

```diff
@@ -76,41 +81,71 @@ func (h *ReqHandler) authorize(req *http.Request, acl []string, alwaysAuth bool,
 	return authorization{}, httpbakery.NewDischargeRequiredErrorForRequest(m, cookiePath, verr, req)
 }
 
-func (h *ReqHandler) checkTerms(id *router.ResolvedURL, req *http.Request) error {
-	user, passwd, err := parseCredentials(req)
-	if err == nil {
-		if user != h.handler.config.AuthUsername || passwd != h.handler.config.AuthPassword {
-			return errgo.WithCausef(nil, params.ErrUnauthorized, "invalid user name or password")
-		}
-		return nil
+func (h *ReqHandler) authorizeWithTerms(req *http.Request, entityId *router.ResolvedURL) (authorization, error) {
+	logger.Infof(
+		"authorize, auth location %q, terms location %q, path: %q, method: %q",
+		h.handler.config.IdentityLocation,
+		h.handler.config.TermsLocation,
+		req.URL.Path,
+		req.Method)
+
+	if entityId == nil {
+		return authorization{}, errgo.WithCausef(nil, params.ErrUnauthorized, "entity id not specified")
 	}
-	entity, err := h.Store.FindEntity(id)
+	entity, err := h.Store.FindEntity(entityId)
 	if err != nil {
-		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+		return authorization{}, errgo.Mask(err, errgo.Is(params.ErrNotFound))
 	}
-	if len(entity.CharmMeta.Terms) > 0 {
-		bk := h.Store.Bakery
-		if h.handler.config.TermsLocation == "" || bk == nil {
-			return errgo.New("charmstore not serving charms requiring agreement to terms and conditions")
-		}
 
-		_, verr := httpbakery.CheckRequest(bk, req, nil, checkers.OperationChecker("get-archive"))
-		if verr == nil {
-			return nil
+	baseEntity, err := h.Store.FindBaseEntity(&entityId.URL, "acls")
+	if err != nil {
+		return authorization{}, errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	if len(entity.CharmMeta.Terms) == 0 {
+		// No need to authenticate if the ACL is open to everyone.
+		for _, name := range baseEntity.ACLs.Read {
+			if name == params.Everyone {
+				return authorization{}, nil
+			}
 		}
-		if _, ok := errgo.Cause(verr).(*bakery.VerificationError); !ok {
-			return errgo.Mask(verr)
+	}
+
+	if len(entity.CharmMeta.Terms) > 0 && h.handler.config.TermsLocation == "" {
+		return authorization{}, errgo.WithCausef(nil, params.ErrUnauthorized, "charmstore not configured to serve charms with terms and conditions")
+	}
+
+	auth, verr := h.checkRequest(req, entityId, opReadArchive)
+	if verr == nil {
+		if err := h.checkACLMembership(auth, baseEntity.ACLs.Read); err != nil {
+			return authorization{}, errgo.WithCausef(err, params.ErrUnauthorized, "")
 		}
-		m, err := bk.NewMacaroon("", nil, []checkers.Caveat{
+		h.auth = auth
+		return auth, nil
+	}
+	if _, ok := errgo.Cause(verr).(*bakery.VerificationError); !ok {
+		return authorization{}, errgo.Mask(verr, errgo.Is(params.ErrUnauthorized))
+	}
+
+	caveats := []checkers.Caveat{
+		checkers.AllowCaveat(opReadArchive, opOther),
```

Comment:
Expiry time of 1day is now added to all macaroons issued by the charmstore (see newMacaroon function), whereas before non of the macaroons (except delegatable) had an expiry time.